canonical/grafana-k8s-operator
# Dashboards links stable

## Issue


If a charm provides a dashboard with an `uid` Grafana will use that uid to create the URL, for instance a dashboard file with:

```json
...
  "title": "Zinc",
  "uid": "z9ghSzQ7z",
  "version": 1
}
```

Grafana will generate a stable URL like: `http://10.1.157.73:3000/d/z9ghSzQ7z/zinc?orgId=1`

The problem appears if the charm provides a dashboad **without** `uid`. 
In this situation Grafana will generate a `uid` based on the JSON file contents... if the JSON file does not change, every time the charm is packed,  refreshed (or deployed and related to Grafana), the URL will be the same.

But if the JSON file is modified (because for example the charm author added a new panel to the dashboard), the URL generated by Grafana will be different.


## Solution


A new private method in `GrafanaDashboardConsumer` was added. This method `_manage_dashboard_uid` will generate an `uid` if the JSON content hasn't.


## Testing Instructions


Run unit tests.

Manually:

- Deploy Grafana
- In [Zinc](https://github.com/jnsgruk/zinc-k8s-operator) charm delete `uid` in Grafana dashboad.
- Pack and deploy the charm
- Check dashboard URL generated by Grafana
- Modify the dashboard file.
- Re-pack and refresh Zinc charm
- Verify the Dashboard URL is the same.
